iib/workers/tasks/opm_operations.py

```diff
@@ -26,31 +30,206 @@
 log = logging.getLogger(__name__)
 
 
-def _gen_port_for_grpc() -> Generator[int, None, None]:
-    """
-    Generate port for gRPC service from range set in IIB config.
+class PortFileLock:
+    """A class representing file-lock used during OPM operations."""
 
-    :raises: IIBError when all ports were already taken
-    """
-    conf = get_worker_config()
-    port_start = conf['iib_grpc_start_port']
-    port_end = port_start + conf['iib_grpc_max_port_tries']
+    def __init__(self, purpose: str, port: int):
+        """
+        Initialize the PortFileLock object.
 
-    port_stack = list(range(port_start, port_end))
-    random.shuffle(port_stack)
+        :param str purpose: Purpose of the lock
+        :param int port: The port number to be locked
+        """
+        self.purpose = purpose
+        self.port = port
+        self.locked = False
+        self.filename = os.path.join(
+            tempfile.gettempdir(),
+            f'iib_{purpose}_{port}.lock',
+        )
 
-    log.debug('Get random ports from range [%d, %d)', port_start, port_end)
+    def lock_acquire(self):
+        """
+        Create a file representing port lock.
 
-    while port_stack:
-        yield port_stack.pop(0)
+        Before trying to create a a port-lock file, we try to check if the port is free.
+        """
+        log.debug(f"Attempt to lock port {self.port}.")
+
+        if self.locked:
+            err_msg = f"Error: Port {self.port} is already locked"
+            log.debug(err_msg)
+            raise IIBError(err_msg)
+
+        # check if the port is free, opm service is doing the check too,
+        # however this way, we do not have to rely on their error message format
+        s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+        try:
+            s.bind(("localhost", self.port))
+        except socket.error:
+            err_msg = f"Port {self.port} is already in use."
+            log.debug(err_msg)
+            raise AddressAlreadyInUse(err_msg)
+        finally:
+            s.close()
+
+        # create file-lock
+        try:
```

Comment:
I think it would be better to combine the locking and testing port in one block of code:
```python
         try:
            s.bind(("localhost", self.port))
            f = os.open(self.filename, os.O_CREAT | os.O_EXCL | os.O_RDWR)
            self.locked = True
            log.debug(f"Port {self.port} used as {self.purpose} was locked.")
        except socket.error:
            err_msg = f"Port {self.port} is already in use."
            log.debug(err_msg)
            raise AddressAlreadyInUse(err_msg)
        except FileExistsError: 
            s.close()
            err_msg = f"Port {self.port} is already locked by other IIB worker."
            log.debug(err_msg)
            raise AddressAlreadyInUse(err_msg)
        finally:
            s.close()
            f.close()
```

Reply:
There seem to be two issues (solved in the last commit) with this solution. 
- socket.error is probably a subclass of FileExistsError, as in this order every time it got caught by socket.error
- we need to check if we opened the file to be able  to close it like that

Please check the last commit, to see the difference. 
I am okay with either solution, which one would you prefer?